src/targets/gpu/hip.cpp

```diff
@@ -77,29 +77,70 @@ void* get_device_ptr(void* hptr)
     return result;
 }
 
-hip_ptr allocate_gpu(std::size_t sz, bool host = false)
+static auto& get_malloc_host_ptrs()
+{
+    static std::unordered_map<void*, std::weak_ptr<void>> cache;
+    return cache;
+}
+
+std::shared_ptr<void> allocate_gpu(std::size_t sz, bool host = false)
 {
     if(sz > get_available_gpu_memory())
         MIGRAPHX_THROW("Memory not available to allocate buffer: " + std::to_string(sz));
-    void* result = nullptr;
-    auto status  = host ? hipHostMalloc(&result, sz) : hipMalloc(&result, sz);
+    void* alloc_ptr = nullptr;
+    auto status     = host ? hipHostMalloc(&alloc_ptr, sz) : hipMalloc(&alloc_ptr, sz);
     if(status != hipSuccess)
     {
         if(host)
             MIGRAPHX_THROW("Gpu allocation failed: " + hip_error(status));
         else
             return allocate_gpu(sz, true);
     }
-    assert(result != nullptr);
-    return hip_ptr{result};
+    assert(alloc_ptr != nullptr);
+    std::shared_ptr<void> result = share(hip_ptr{alloc_ptr});
+    if(host)
+    {
+        get_malloc_host_ptrs()[alloc_ptr] = result;
```

Comment:
This isnt thread-safe.

Reply:
Since this is modified and read in two place we are going to need a class so the mutex stays with the map:

```cpp
struct host_ptr_cache
{
    std::unordered_map<void*, std::weak_ptr<void>> cache
    std::mutex m;

    std::shared_ptr<void> get(void* p)
    {
        std::lock_guard<std::mutex> lock(m);
        auto it = cache.find(ptr);
        if(it != cache.end())
            return it->second.lock();
        return nullptr;
    }

    void put(std::shared_ptr<void> p)
    {
        std::lock_guard<std::mutex> lock(m);
        cache[p.get()] = p;
    }
};

host_ptr_cache& get_host_ptr_cache()
{
    static host_ptr_cache cache;
    return cache;
}
```